services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,261 @@
+
+        messageProducer.send(topic, message, new ListenableFutureCallback<SendResult<String, Message>>() {
+            @Override
+            public void onSuccess(SendResult<String, Message> result) {
+                logger.debug("Message sent: topic={}, message={}", topic, message);
+
+                pendingChunkedRequests.put(message.getCorrelationId(), future);
+                messagesChains.put(message.getCorrelationId(), new ArrayList<>());
+            }
+
+            @Override
+            public void onFailure(Throwable exception) {
+                logger.error("Unable to send message: topic={}, message={}", topic, message, exception);
+                future.completeExceptionally(exception);
+            }
+        });
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     *
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Received invalid message: {}", message);
+            return;
+        }
+
+        if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+
+            completeWithError(error);
+        }
+
+        if (message instanceof InfoMessage) {
+            if (message instanceof ChunkedInfoMessage) {
+                processChunkedMessage((ChunkedInfoMessage) message);
+            } else if (pendingRequests.containsKey(message.getCorrelationId())) {
+                InfoMessage infoMessage = (InfoMessage) message;
+                pendingRequests.remove(message.getCorrelationId())
+                        .complete(infoMessage.getData());
+            }
+        }
+    }
+
+    /**
+     * Performs searching and collecting all chunked messages into one chain if possible.
+     */
+    private synchronized void processChunkedMessage(ChunkedInfoMessage received) {
+        unlinkedChunks.put(received.getCorrelationId(), received);
+
+        List<ChunkedInfoMessage> chain = messagesChains.remove(received.getCorrelationId());
+        if (chain != null) {
+            String nextRequestId = received.getCorrelationId();
+
+            // adds already received messages to the current chain.
+            do {
+                ChunkedInfoMessage message = unlinkedChunks.remove(nextRequestId);
+                if (message.getData() != null) {
```

Comment:
If you move this filter inside `completeChain` you can drop `chain.isEmpty()` check... because chaincorrelation id will be always in first chain element.